be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala

```diff
@@ -42,66 +51,86 @@ object GetMessagesByIdResponseHandler extends AskPatternConstants {
     case value @ _ => value
   }
 
-  // This function will try to digest the set of messages for an entire channel
-  // It will go channel by channel until it finished processing messages or MAX_RETRY is reached
+  /** Will try to digest each GraphMessage until their retry-counter reaches 0 or they all get validated
+    *
+    * @param receivedResponse
+    *   The wrapped messages with their counter
+    * @param validatorFlow
+    *   The validator to push the messages through
+    * @param system
+    *   Implicit actor system to use the given validator
+    * @return
+    *   true if all messages could be validated, otherwise false
+    */
   @tailrec
   private def passThroughPipeline(
-      receivedResponse: Map[Channel, Set[GraphMessage]],
-      validatorFlow: Flow[GraphMessage, GraphMessage, NotUsed],
-      remainingAttempts: Int
+      receivedResponse: List[(GraphMessage, Int)],
+      validatorFlow: Flow[GraphMessage, GraphMessage, NotUsed]
   )(implicit system: ActorSystem): Boolean = {
-    if (receivedResponse.isEmpty || remainingAttempts <= 0) {
+    if (receivedResponse.isEmpty || receivedResponse.forall(elem => elem._2 <= 0)) {
```

Comment:
The second check `receivedResponse.forall(elem => elem._2 <= 0)` is actually not really necesssary as messages with `retry == 0` are ignored in the following `foreach(...)`, so if all the responses received have 0 retry left, the next `passThroughPipeline` will have an empty list of responses and will therefor still return (so it would avoid iterating over the entire list of responses 2 times per call instead of only 1)